src/batchproof_container.cpp

```diff
@@ -159,39 +160,71 @@ void BatchProofContainer::erase(std::vector<LelantusSigmaProofData>* vProofs, co
-        sigmaState->GetAnonymitySet(
-                itr.first.first,
-                itr.first.second.first,
-                itr.first.second.second,
-                anonymity_set);
-
-        size_t m = itr.second.size();
-        std::vector<Scalar> serials;
-        serials.reserve(m);
-        std::vector<bool> fPadding;
-        fPadding.reserve(m);
-        std::vector<size_t> setSizes;
-        setSizes.reserve(m);
-        std::vector<sigma::SigmaPlusProof<Scalar, GroupElement>> proofs;
-        proofs.reserve(m);
-
-        for (auto& proofData : itr.second) {
-            serials.emplace_back(proofData.coinSerialNumber);
-            fPadding.emplace_back(proofData.fPadding);
-            setSizes.emplace_back(proofData.anonymitySetSize);
-            proofs.emplace_back(proofData.sigmaProof);
-        }
+    else
+        return;
+    std::size_t threadsMaxCount = std::min((unsigned int)sigmaProofs.size(), boost::thread::hardware_concurrency());
+    std::vector<boost::future<bool>> parallelTasks;
+    parallelTasks.reserve(threadsMaxCount);
+    ParallelOpThreadPool<bool> threadPool(threadsMaxCount);
+
+    auto params = sigma::Params::get_default();
+    sigma::SigmaPlusVerifier<Scalar, GroupElement> sigmaVerifier(params->get_g(), params->get_h(), params->get_n(), params->get_m());
+
+    auto itr = sigmaProofs.begin();
+    for (std::size_t j = 0; j < sigmaProofs.size(); j += threadsMaxCount) {
+        for (std::size_t i = j; i < j + threadsMaxCount; ++i) {
+            if (i < sigmaProofs.size()) {
+                std::vector<GroupElement> anonymity_set;
+                sigma::CSigmaState* sigmaState = sigma::CSigmaState::GetState();
+                sigmaState->GetAnonymitySet(
+                        itr->first.first,
+                        itr->first.second.first,
+                        itr->first.second.second,
+                        anonymity_set);
+
+                size_t m = itr->second.size();
+                std::vector<Scalar> serials;
+                serials.reserve(m);
+                std::vector<bool> fPadding;
+                fPadding.reserve(m);
+                std::vector<size_t> setSizes;
+                setSizes.reserve(m);
+                std::vector<sigma::SigmaPlusProof<Scalar, GroupElement>> proofs;
+                proofs.reserve(m);
+
+                for (auto& proofData : itr->second) {
+                    serials.emplace_back(proofData.coinSerialNumber);
+                    fPadding.emplace_back(proofData.fPadding);
+                    setSizes.emplace_back(proofData.anonymitySetSize);
+                    proofs.emplace_back(proofData.sigmaProof);
+                }
 
-        auto params = sigma::Params::get_default();
-        sigma::SigmaPlusVerifier<Scalar, GroupElement> sigmaVerifier(params->get_g(), params->get_h(), params->get_n(), params->get_m());
+                parallelTasks.emplace_back(threadPool.PostTask([=]() {
+                    DoNotDisturb dnd;
```

Comment:
dnd should be in the same function frame that gets `std::future` result, not in lambda

Reply:
done